### PR TITLE
[FrameworkBundle] Deprecate auto-injection of the container in AbstractController instances

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Should enhance DX by preventing situations like #27436.

### CODE REVIEW

**Review by ogizanagi — APPROVED** [review]

**Review by curry684 — APPROVED** [review]
Fair solution for the main underlying confusing issue.

**Review by chalasr — APPROVED** [review]

**Review by 20uf — APPROVED** [review]

**Review by xabbuh — APPROVED** [review]

**Review by Tobion — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerResolver.php [hunk 1]
```diff
@@ -51,16 +51,23 @@ protected function createController($controller)
      */
     protected function instantiateController($class)
     {
-        return $this->configureController(parent::instantiateController($class));
+        return $this->configureController(parent::instantiateController($class), $class);
     }
 
-    private function configureController($controller)
+    private function configureController($controller, $class)
```
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/Controller/ControllerResolver.php]: `string $class` ?

**Review by dunglas — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerResolver.php [hunk 2]
```diff
@@ -51,16 +51,23 @@ protected function createController($controller)
...
     {
         if ($controller instanceof ContainerAwareInterface) {
             $controller->setContainer($this->container);
         }
-        if ($controller instanceof AbstractController && null !== $previousContainer = $controller->setContainer($this->container)) {
-            $controller->setContainer($previousContainer);
+        if ($controller instanceof AbstractController) {
+            $previousContainer = $controller->setContainer($this->container);
+            if (null !== $previousContainer) {
```
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/Controller/ControllerResolver.php]: This test can maybe be inverted `(if null ===) else` for clarity?

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerResolver.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Controller/ControllerResolver.php]: updated

**Pull request merged**